zerver/lib/user_groups.py

```diff
@@ -194,6 +194,49 @@ def access_user_group_for_update(
     raise JsonableError(_("Insufficient permission"))
 
 
+def check_unused_anonymous_user_groups(realm: Realm) -> set[int]:
+    anonymous_group_ids = UserGroup.objects.filter(realm=realm, named_user_group=None).values_list(
+        "id", flat=True
+    )
+
+    realm_setting_group_ids = [
+        getattr(realm, setting_name + "_id")
+        for setting_name in Realm.REALM_PERMISSION_GROUP_SETTINGS
+    ]
+
+    stream_settings_list = [
+        setting_name + "_id" for setting_name in Stream.stream_permission_group_settings
+    ]
+    streams = Stream.objects.filter(realm=realm).values(*stream_settings_list)
```

Comment:
What if we add filters for both stream and group settings instead of fetching these values in Python?

Something like 
```
anonymous_groups = UserGroup.objects.filter(realm=realm, named_user_group=None)
for setting_name in Stream.stream_permission_group_settings:
        anonymous_groups = anonymous_groups.filter(
            ~Exists(
                Stream.objects.filter(
                    realm=realm,
                    **{f"{setting_name}_id": OuterRef("id")},
                )
            )
        )
```

We add a similar filter for groups as well and resolve it at the end so we are not fetching streams and groups separately + storing them in Python. What do you think?